### ISSUE TITLE
How to set event handlers in modules, that depend one to another?

### ISSUE THREAD

**Issue by yujin1st** [yiisoft/yii2#11624]
There are two modules that handle events of each others.
Where should handlers initialization be placed? 
You get a recursion, if you place it in init section of module.

 ```php
//module 1
public function init() {
    parent::init();
    if ($app->hasModule('module2')) {
      $module = $app->getModule('module2');
      $module->on('someModule2Event', function ($event) {
        // do something
      });
  }
 //  the same in module 2
```

**Comment by mdmunir** [comment]
```php
// module 1
public function init() {
    parent::init();
    $this->on('someModule1Event', function($event){
        // call module2 method
    });
  }
```

**Comment by yujin1st** [comment]
@mdmunir, i need to setup event handler in first module for 2nd one  and vise-versa. I don't need to process own modules events.

**Comment by unclead** [comment]
You can use Application level events instead of Module level

```
public function init() {
    parent::init();
    Yii::$app->on('someModule2Event', function($event){
        ...
    });
  }

```

**Comment by mdmunir** [comment]
@yujin1st 
simple, use forum for asking.

**Comment by yujin1st** [comment]
@unclead, what if have many modules with many event - it will easy to mess with it. That's why i want to use module level.

@mdmunir, let me rephrase it as framework issue:
1. Make 2 modules, that has a couple events. Module 1 should listen events of 2nd module, and vise-versa. 
2. Put it in bootstrap section of config so we can init hadlers
3. Write code in both modules: 
```php
//module1
public function init() {
    parent::init();
    if ($app->hasModule('module2')) {
      $module = $app->getModule('module2');
      $module->on('someModule2Event', function ($event) {
        // do something
      });
  }

//module2
public function init() {
    parent::init();
    if ($app->hasModule('module2')) {
      $module = $app->getModule('module2');
      $module->on('someModule2Event', function ($event) {
        // do something
      });
  }
```php

**Comment by mdmunir** [comment]
@yujin1st I undestand what are you talking. But, first and all. Its not framework issue, its your issue. You that make it recursive.
So, you must change your strategy. I give you a clue, use class level event instead of instance level event.

Last, i dont realy understand why you using `module` as event handler. Why not using `BootstrapInterface` or `Behavior` or something else? Do you understand what module are?

**Comment by yujin1st** [comment]
@mdmunir,
I have a cms with modules (user, content, catalog, cart, photo, others) that provide many event for proccessing it's data. Now It uses app level global events, modules are hardcored in config, and in new version i want to to make modules dynamical and independent - you can turn off something, install new modules - and system should work. And i want remove intermediate component for event management.
